src/controller/project/controller.go

```diff
@@ -128,91 +161,174 @@ func (c *controller) GetByName(ctx context.Context, projectName string, options
 		return nil, err
 	}
 
-	opts := newOptions(options...)
-	if opts.WithOwner {
-		if err := c.loadOwners(ctx, models.Projects{p}); err != nil {
-			return nil, err
-		}
+	if err := c.assembleProjects(ctx, models.Projects{p}, options...); err != nil {
+		return nil, err
 	}
-	return c.assembleProject(ctx, p, newOptions(options...))
+
+	return p, nil
 }
 
-func (c *controller) List(ctx context.Context, query *models.ProjectQueryParam, options ...Option) ([]*models.Project, error) {
+func (c *controller) List(ctx context.Context, query *q.Query, options ...Option) ([]*models.Project, error) {
 	projects, err := c.projectMgr.List(ctx, query)
 	if err != nil {
 		return nil, err
 	}
 
-	opts := newOptions(options...)
-	if opts.WithOwner {
-		if err := c.loadOwners(ctx, projects); err != nil {
-			return nil, err
-		}
+	if len(projects) == 0 {
+		return projects, nil
 	}
 
-	for _, p := range projects {
-		if _, err := c.assembleProject(ctx, p, opts); err != nil {
-			return nil, err
-		}
+	if err := c.assembleProjects(ctx, projects, options...); err != nil {
+		return nil, err
 	}
 
 	return projects, nil
 }
 
-func (c *controller) loadOwners(ctx context.Context, projects models.Projects) error {
-	owners, err := c.userMgr.List(ctx, q.New(q.KeyWords{"user_id__in": projects.OwnerIDs()}))
-	if err != nil {
-		return err
+func (c *controller) Update(ctx context.Context, p *models.Project) error {
```

Comment:
the actions of update are in a transaction, right?

Reply:
I don't think so because the manager of the CVE allowlist still uses global ormer.

Reply:
so, it could happen that, allowlist is updated success, but others are fail?

Reply:
Yes, this scenario may happen. Emm, update metadata first may resolve this issue. I will change the order to update the metadata first so that when the allowlist update failed the changes for the metadata can be rollback.